### ISSUE TITLE
Non-negative matrix factorization using Autograd

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4070]
Non-negative matrix factorization using Autograd<br>
<br>
http://bit.ly/2VU3V7r<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<div>
<div>
<p>In a <a href="https://nipunbatra.github.io/blog/2017/nnmf-tensorflow.html">previous post</a>, we had seen how to perfom non-negative matrix factorization (NNMF) using Tensorflow. In this post, we will look at performing NNMF using <a href="https://github.com/HIPS/autograd">Autograd</a>. Like Tensorflow, Autograd allows automatic gradient calculation.</p>
</div>
</div>
<div>
<div>
<p>In&nbsp;[2]:</p>
<div>
<div>
<div>
<pre>
<span>import</span> <span>autograd.numpy</span> <span>as</span> <span>np</span>
<span>import</span> <span>pandas</span> <span>as</span> <span>pd</span>
<span>import</span> <span>matplotlib.pyplot</span> <span>as</span> <span>plt</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Creating the matrix to be decomposed<a href="#Creating-the-matrix-to-be-decomposed">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[3]:</p>
<div>
<div>
<div>
<pre>
<span>A</span> <span>=</span> <span>np</span><span>.</span><span>array</span><span>([[</span><span>3</span><span>,</span> <span>4</span><span>,</span> <span>5</span><span>,</span> <span>2</span><span>],</span>
                   <span>[</span><span>4</span><span>,</span> <span>4</span><span>,</span> <span>3</span><span>,</span> <span>3</span><span>],</span>
                   <span>[</span><span>5</span><span>,</span> <span>5</span><span>,</span> <span>4</span><span>,</span> <span>3</span><span>]],</span> <span>dtype</span><span>=</span><span>np</span><span>.</span><span>float32</span><span>)</span><span>.</span><span>T</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<div>
<p>Out[5]:</p>
<div>
<pre>
array([[ nan,   4.,   5.],
       [  4.,   4.,   5.],
       [  5.,   3.,   4.],
       [  2.,   3.,   3.]], dtype=float32)
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Defining the cost function<a href="#Defining-the-cost-function">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[210]:</p>
<div>
<div>
<div>
<pre>
<span>def</span> <span>cost</span><span>(</span><span>W</span><span>,</span> <span>H</span><span>,</span> <span>la</span> <span>=</span> <span>.</span><span>05</span><span>):</span>
    <span>pred</span> <span>=</span> <span>np</span><span>.</span><span>dot</span><span>(</span><span>W</span><span>,</span> <span>H</span><span>)</span>
    <span>mask</span> <span>=</span> <span>~</span><span>np</span><span>.</span><span>isnan</span><span>(</span><span>A</span><span>)</span>
    <span>C</span> <span>=</span> <span>np</span><span>.</span><span>sqrt</span><span>(((</span><span>pred</span> <span>-</span> <span>A</span><span>)[</span><span>mask</span><span>]</span><span>.</span><span>flatten</span><span>()</span> <span>**</span> <span>2</span><span>)</span><span>.</span><span>mean</span><span>(</span><span>axis</span><span>=</span><span>None</span><span>))</span>
    <span>D</span> <span>=</span> <span>la</span><span>*</span><span>len</span><span>(</span><span>H</span><span>[</span><span>H</span><span>&gt;</span><span>0</span><span>])</span>
    <span>#D = np.linalg.norm(H, ord=2)</span>
    <span>D</span> <span>=</span> <span>max</span><span>(</span><span>np</span><span>.</span><span>sum</span><span>(</span><span>np</span><span>.</span><span>abs</span><span>(</span><span>H</span><span>),</span> <span>axis</span><span>=</span><span>0</span><span>))</span>
    <span>#print(C, D)</span>
    <span>return</span> <span>D</span> <span>+</span><span>C</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[222]:</p>
<div>
<div>
<div>
<pre>
<span>rank</span> <span>=</span> <span>5</span>
<span>learning_rate</span><span>=</span><span>0.5</span>
<span>n_steps</span> <span>=</span> <span>4000</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Gradient of cost wrt params W and H<a href="#Gradient-of-cost-wrt-params-W-and-H">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[223]:</p>
<div>
<div>
<div>
<pre>
<span>from</span> <span>autograd</span> <span>import</span> <span>grad</span><span>,</span> <span>multigrad</span>
<span>grad_cost</span><span>=</span> <span>multigrad</span><span>(</span><span>cost</span><span>,</span> <span>argnums</span><span>=</span><span>[</span><span>0</span><span>,</span><span>1</span><span>])</span>
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Main gradient descent routine<a href="#Main-gradient-descent-routine">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[224]:</p>
<div>
<div>
<div>
<pre>
<span>shape</span> <span>=</span> <span>A</span><span>.</span><span>shape</span>
<span>H</span> <span>=</span>  <span>np</span><span>.</span><span>abs</span><span>(</span><span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>(</span><span>rank</span><span>,</span> <span>shape</span><span>[</span><span>1</span><span>]))</span>
<span>W</span> <span>=</span>  <span>np</span><span>.</span><span>abs</span><span>(</span><span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>(</span><span>shape</span><span>[</span><span>0</span><span>],</span> <span>rank</span><span>))</span>
<span>print</span> <span>"Iteration, Cost"</span>
<span>for</span> <span>i</span> <span>in</span> <span>range</span><span>(</span><span>n_steps</span><span>):</span>
    
    <span>if</span> <span>i</span><span>%</span><span>500</span>==0:
        <span>print</span> <span>"*"</span><span>*</span><span>20</span>
        <span>print</span> <span>i</span><span>,</span><span>","</span><span>,</span> <span>cost</span><span>(</span><span>W</span><span>,</span> <span>H</span><span>)</span>
    <span>del_W</span><span>,</span> <span>del_H</span> <span>=</span> <span>grad_cost</span><span>(</span><span>W</span><span>,</span> <span>H</span><span>)</span>
    <span>W</span> <span>=</span>  <span>W</span><span>-</span><span>del_W</span><span>*</span><span>learning_rate</span>
    <span>H</span> <span>=</span>  <span>H</span><span>-</span><span>del_H</span><span>*</span><span>learning_rate</span>
    
    <span># Ensuring that W, H remain non-negative. This is also called projected gradient descent</span>
    <span>W</span><span>[</span><span>W</span><span>&lt;</span><span>0</span><span>]</span> <span>=</span> <span>0</span>
    <span>H</span><span>[</span><span>H</span><span>&lt;</span><span>0</span><span>]</span> <span>=</span> <span>0</span>
</pre>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<div>
<pre>
Iteration, Cost
********************
0 , 11.1645842634
********************
500 , 3.18111432175
********************
1000 , 3.16056906983
********************
1500 , 3.1625775801
********************
2000 , 3.13009606497
********************
2500 , 3.26063346835
********************
3000 , 3.00275139422
********************
3500 , 3.17581052995
</pre>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<div>
<p>Out[225]:</p>
<div>
<div>
<table border="1">
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>0.00000</td>
<td>0.00000</td>
<td>0.000000</td>
<td>3.838099</td>
<td>0.00000</td>
</tr>
<tr>
<th>1</th>
<td>0.00000</td>
<td>0.00000</td>
<td>0.000000</td>
<td>3.747767</td>
<td>0.00000</td>
</tr>
<tr>
<th>2</th>
<td>0.00000</td>
<td>0.00000</td>
<td>1.483943</td>
<td>3.358283</td>
<td>0.00000</td>
</tr>
<tr>
<th>3</th>
<td>0.87381</td>
<td>1.72481</td>
<td>0.000000</td>
<td>2.104210</td>
<td>0.41053</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<div>
<p>Out[227]:</p>
<div>
<div>
<table border="1">
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>0.000000</td>
<td>0.058065</td>
<td>0.000000</td>
</tr>
<tr>
<th>1</th>
<td>0.000000</td>
<td>0.114614</td>
<td>0.000000</td>
</tr>
<tr>
<th>2</th>
<td>0.336290</td>
<td>0.075604</td>
<td>0.000000</td>
</tr>
<tr>
<th>3</th>
<td>1.152227</td>
<td>1.291911</td>
<td>0.669434</td>
</tr>
<tr>
<th>4</th>
<td>0.000000</td>
<td>0.027280</td>
<td>0.000000</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<p>In&nbsp;[228]:</p>
<div>
<div>
<div>
<pre>
<span>pred</span> <span>=</span> <span>np</span><span>.</span><span>dot</span><span>(</span><span>W</span><span>,</span> <span>H</span><span>)</span>
<span>pred_df</span> <span>=</span> <span>pd</span><span>.</span><span>DataFrame</span><span>(</span><span>pred</span><span>)</span><span>.</span><span>round</span><span>()</span>
<span>pred_df</span>
</pre>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<p>Out[228]:</p>
<div>
<div>
<table border="1">
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>4.0</td>
<td>5.0</td>
<td>3.0</td>
</tr>
<tr>
<th>1</th>
<td>4.0</td>
<td>5.0</td>
<td>3.0</td>
</tr>
<tr>
<th>2</th>
<td>4.0</td>
<td>4.0</td>
<td>2.0</td>
</tr>
<tr>
<th>3</th>
<td>2.0</td>
<td>3.0</td>
<td>1.0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
<div>
<div>
<div>
<div>
<p>Out[229]:</p>
<div>
<div>
<table border="1">
<thead>
<tr>
<th></th>
<th>0</th>
<th>1</th>
<th>2</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>NaN</td>
<td>4.0</td>
<td>5.0</td>
</tr>
<tr>
<th>1</th>
<td>4.0</td>
<td>4.0</td>
<td>5.0</td>
</tr>
<tr>
<th>2</th>
<td>5.0</td>
<td>3.0</td>
<td>4.0</td>
</tr>
<tr>
<th>3</th>
<td>2.0</td>
<td>3.0</td>
<td>3.0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via nipunbatra.github.io http://bit.ly/2Eq3lnf<br>
<br>
May 23, 2019 at 02:35PM
